feat(shard-distributor): record a smoothed per shard load in etcd

Theis-Mathiassen opened the pull request:
<!-- Describe what has changed in this PR -->
**What changed?**
We added functionality to record the load as a moving average for each shard, where the weight of a new data point depends on how recently the average was last updated.

Furthermore it is now using cache, to reduce the read intensity on etcd.

<!-- Tell your future self why have you made these changes -->
**Why?**
This is done to smooth the load input for the shard distributor, this is desirable as the load can change sporadically.
It is also necessary to save the load of each shard in ETCD, as to persist it (In case the handler crashes) and make it available to each instance of shard distributors.

<!-- How have you verified this change? Tested locally? Added a unit test? Checked in staging env? -->
**How did you test it?**
We have created some unit tests, and tried to run it with the canary service:
TestRecordHeartbeatUpdatesShardStatistics:
This test verifies that when an executor sends a heartbeat with ShardLoad information for a shard, the ShardStatistics for that shard are correctly updated in the store, specifically the SmoothedLoad and LastUpdateTime. It also ensures that LastMoveTime remains unchanged if not explicitly updated.

TestRecordHeartbeatSkipsShardStatisticsWithNilReport:
This test ensures that if an executor's heartbeat includes a nil ShardStatusReport for a particular shard, the existing ShardStatistics for that shard are not updated or created. It also verifies that valid shard reports are processed correctly.

These tests were run with the command: ```ETCD=1 go test ./service/sharddistributor/...```

<!-- Assuming the worst case, what can be broken when deploying this change to production? -->
**Potential risks**
None, since it is for the shard distributor, which is not utilized in production yet.

<!-- Is it notable for release? e.g. schema updates, configuration or data migration required? If so, please mention it, and also update CHANGELOG.md -->
**Release notes**
It is not, since it is for the shard distributor, which is not utilized in production yet.

<!-- Is there any documentation updates should be made for config, https://cadenceworkflow.io/docs/operation-guide/setup/ ? If so, please open an PR in https://github.com/cadence-workflow/cadence-docs -->
**Documentation Changes**
No, but maybe some documentation should be created, later.

----
## Summary by Gitar

- **Test configuration updates:**
  - Updated test migration mode from `DISTRIBUTEDPASSTHROUGH` to `LOCALPASSTHROUGH` in `processor_test.go` to reflect current testing requirements.

<sub>This will update automatically on new commits.</sub>

AndreasHolt committed:
feat(shard distributor): add shard key helpers and metrics state

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
feat(shard distributor): persist shard metrics in etcd store

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
feat(shard distributor): persist shard metrics in etcd store

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>
Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
fix(shard distributor): update LastMoveTime in the case where a shard is being reassigned in AssignShard

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
test(shard distributor): add tests for shard metrics

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix(shard distributor): modify comment

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix(shard distributor): add atomic check to prevent metrics race

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix(shard distributor): apply shard metric updates in a second phase to not overload etcd's 128 max ops per txn

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
feat(shard distributor): move shard metric updates out of AssignShards txn and retry monotonically

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>
Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
feat(shard distributor): move shard metric updates out of AssignShards txn and retry monotonically

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix(shard distributor): keep NamespaceState revisions tied to assignments

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix(shard distributor): keep NamespaceState revisions tied to assignments

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>
Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
refactor(shard distributor): use shard cache and clock for preparing shard metrics, move out to staging to separate function

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
refactor(shard distributor): use shard cache and clock for preparing shard metrics, move out to staging to separate function

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>
Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
test(shard distributor): BuildShardPrefix, BuildShardKey, ParseShardKey

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
feat(shard distributor): simplify shard metrics updates

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
feat(shard distributor): simplify shard metrics updates

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>
Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
refactor(shard distributor): ShardMetrics renamed to ShardStatistics. And more idiomatic naming of collection vs singular type

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
refactor(shard distributor): ShardMetrics renamed to ShardStatistics. And more idiomatic naming of collection vs singular type

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>
Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
test(shard distributor): small changes to shard key tests s.t. they look more like executor key tests

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix(shard distributor): no longer check for key type ShardStatisticsKey in BuildShardKey, as we don't use it

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
refactor(shard distributor): found a place where I forgot to rename to "statistics"

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
refactor(shard distributor): found a place where I forgot to rename to "statistics"

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>
Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
feat: function to update shard statistics from heartbeat (currently no ewma)

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>
Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
feat: function to update shard statistics from heartbeat (currently no ewma)

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
test(shard distributor): add tests to verify statistics are updated at heartbeat

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>
Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
test(shard distributor): add tests to verify statistics are updated at heartbeat

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix(shard distributor): move non-exported helpers to end of file to follow conventions

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>
Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
fix(shard distributor): move non-exported helpers to end of file to follow conventions

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
feat(shard distributor): calculate smoothed load (ewma) using the ShardStatistics

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>
Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
feat(shard distributor): calculate smoothed load (ewma) using the ShardStatistics

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix(shard distributor): log invalid shard load

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>
Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
fix(shard distributor): log invalid shard load

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
feat(shard distributor): clean up the shard statistics

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
test(shard distributor): add test case for when shard stats are deleted

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>
Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
test(shard distributor): add test case for when shard stats are deleted

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix(shard distributor): add mapping (new metric)

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
feat(shard distributor): retain shard stats while shards are within heartbeat TTL

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>
Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
feat(shard distributor): retain shard stats while shards are within heartbeat TTL

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

Theis-Mathiassen committed:
chore: added logger warning and simplified ewma calculation

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
chore: added logger warning and simplified ewma calculation

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

AndreasHolt committed:
fix: remove duplicate test introduced in merge

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>
Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
fix: remove duplicate test introduced in merge

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

Theis-Mathiassen committed:
chore: consistent error checking, and rename function

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
chore: consistent error checking, and rename function

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
chore: added decompress to unmarshal

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
chore: added decompress to unmarshal

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
chore: removed an old struct that appeared during rebase

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen commented on service/sharddistributor/store/etcd/executorstore/etcdstore.go:
```diff
@@ -153,9 +154,108 @@ func (s *executorStoreImpl) RecordHeartbeat(ctx context.Context, namespace, exec
+			continue
+		}
+
+		_, err = s.client.Put(ctx, shardStatsKey, string(payload))
```
Yeah I think that is possible, so as an example if load from heartbeat is within +/- 5% of the currently stored value, we do not update it?

That should work, however because of how ewmaSmoothedLoad calculates the new load, if there have been a longer time since last load change, the new load will have greater impact.

So I fear if we have a very stable shard load with some spikes in the load, the smoothing effect is lost, excuse my drawing, but i hope it illustrates the idea:
<img width="1116" height="1027" alt="image" src="https://github.com/user-attachments/assets/c03fd912-e647-426d-b410-7568883d8e13" />

It might still be possible, we will probably just have to rethink the update function, how we can avoid this.

AndreasHolt replied:
Have implemented something that can help throttle the writes in 158e0307b

AndreasHolt committed:
feat(shard distributor): throttle shard-stat writes

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
Merge branch 'heartbeat-shard-statistics' of github.com:AndreasHolt/cadencefork into heartbeat-shard-statistics

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix(shard distributor): linter error

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
feat(shard distributor): decouple shard stats write-throttling decision from heartbeat TTL

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
Merge branch 'master' into heartbeat-shard-statistics

AndreasHolt committed:
fix(shard-distributor):  inverted condition in shard stats cleanup loop

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

Theis-Mathiassen committed:
chore(shard-distributor): did some formatting, and use current load instead of new ewma for determinig if it should persist

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

AndreasHolt committed:
Merge branch 'master' into heartbeat-shard-statistics

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix(shard distributor): decouple shard assignment from stats writes

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

eleonoradgr commented on service/sharddistributor/store/etcd/executorstore/etcdstore.go:
```diff
@@ -165,9 +178,124 @@ func (s *executorStoreImpl) RecordHeartbeat(ctx context.Context, namespace, exec
 		return fmt.Errorf("record heartbeat: %w", err)
 	}
+
+	s.recordShardStatistics(ctx, namespace, executorID, request.ReportedShards)
```
Let's discuss if this should happen syncrhronously when we record heartbeat

AndreasHolt committed:
feat: add new ExecutorKeyType for stats, add case to GetState and remove old fetching logic of shard-level stats

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
feat: update etcdstore.go to support new way of storing stats

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
test: persistence of new stats

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
test: update test, remove batching

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
chore: ExecutorShardStatisticsKey is not significant

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix: linter

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
chore: delete shard key related helpers and tests

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
chore: add logs to cleanup

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix: comment

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
chore: remove debug logs

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
CI rerun

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix: remove "len(ops) == 0" check

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

Theis-Mathiassen committed:
Merge remote-tracking branch 'origin/stats-etcd-refactor' into heartbeat-shard-statistics

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

Theis-Mathiassen committed:
feat: changed heartbeat record statistics in new format

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

AndreasHolt committed:
fix: potential fix to smothed load not working as expected

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

Theis-Mathiassen committed:
feat: implemented getting shard statistics, using cache

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
test: added and altered tests to work with statistics from cache

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
Merge branch 'heartbeat-shard-statistics' of github.com:AndreasHolt/cadence into heartbeat-shard-statistics

AndreasHolt committed:
feat: locks and optimization

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
chore: comments

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

Theis-Mathiassen committed:
chore: added comment describing a complicated function

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

AndreasHolt committed:
Merge branch 'heartbeat-shard-statistics' of github.com:AndreasHolt/cadence into heartbeat-shard-statistics

Theis-Mathiassen committed:
Merge remote-tracking branch 'origin/master' into heartbeat-shard-statistics

Theis-Mathiassen committed:
chore: refactored a function, only calling 1 other function

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

AndreasHolt committed:
chore: set LastMoveTime to zero for new shards

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
Merge branches 'heartbeat-shard-statistics' and 'heartbeat-shard-statistics' of github.com:AndreasHolt/cadence into heartbeat-shard-statistics

AndreasHolt committed:
fix: event type

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

Theis-Mathiassen committed:
chore: removed unused test helper functions causing build errors

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

AndreasHolt committed:
chore: consistent naming

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
Merge branch 'heartbeat-shard-statistics' of github.com:AndreasHolt/cadence into heartbeat-shard-statistics

dkrotx commented on service/sharddistributor/store/etcd/executorstore/etcdstore.go:
```diff
@@ -136,6 +137,74 @@ func (s *executorStoreImpl) RecordHeartbeat(ctx context.Context, namespace, exec
+		load := report.ShardLoad
+		if math.IsNaN(load) || math.IsInf(load, 0) {
+			s.logger.Warn(
+				"invalid shard load reported; skipping EWMA update",
```
EWMA?

Theis-Mathiassen replied:
We abbreviated Exponential weighted moving average.
It is renamed here to make it more clear: [8938199](https://github.com/cadence-workflow/cadence/commit/893819970555c893455de9b5afd050eeea738b47)
We also renamed the function, here: [fef9cb8](https://github.com/cadence-workflow/cadence/commit/fef9cb8c0044930d2d540599bb5da2a4887eb3fb)

dkrotx commented on service/sharddistributor/store/etcd/executorstore/etcdstore.go:
```diff
@@ -136,6 +137,74 @@ func (s *executorStoreImpl) RecordHeartbeat(ctx context.Context, namespace, exec
+				tag.ShardExecutor(executorID),
+				tag.ShardKey(shardID),
+			)
+			continue
```
This part is more like converting. Shoult it even be here?

dkrotx commented on service/sharddistributor/store/etcd/executorstore/etcdstore.go:
```diff
@@ -136,6 +137,74 @@ func (s *executorStoreImpl) RecordHeartbeat(ctx context.Context, namespace, exec
+		stats.SmoothedLoad = newSmoothed
+		stats.LastUpdateTime = etcdtypes.Time(now)
+
+		statsUpdate.stats[shardID] = stats
```
I would extract this calculation from etcdstore.go as it has nothing to do with storage/model.

Theis-Mathiassen replied:
That makes sense, we moved it to its separate function here: [4f649dd](https://github.com/cadence-workflow/cadence/commit/4f649dd91135324523aaca2689b5b0775f747667)

dkrotx commented on service/sharddistributor/store/etcd/executorstore/etcdstore.go:
```diff
@@ -735,44 +807,63 @@ func (s *executorStoreImpl) prepareShardStatisticsUpdates(ctx context.Context, n
 			} else {
-				stats.SmoothedLoad = 0
-				stats.LastUpdateTime = etcdtypes.Time(now)
+				shardStatToMove.LastMoveTime = etcdtypes.Time(now)
```
nit: can we combine this with moving? Then it will be more clear why the field is updted here, and not updated in `if()`

Theis-Mathiassen replied:
That should be fixed in [aa9553e](https://github.com/cadence-workflow/cadence/commit/aa9553ebe946ecfd8f3633ed683a3cabd9284afc).

dkrotx commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -97,6 +100,42 @@ func (n *namespaceShardToExecutor) GetExecutorModRevisionCmp() ([]clientv3.Cmp,
+	stats := make(map[string]etcdtypes.ShardStatistics)
+	if len(resp.Kvs) == 0 {
+		return stats, nil
+	}
```
this should be after retrieving resp/analyzing error. Otherwise it looks related to the previous line (`stats`).

dkrotx commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -97,6 +100,42 @@ func (n *namespaceShardToExecutor) GetExecutorModRevisionCmp() ([]clientv3.Cmp,
+
+	n.executorStatisticsLock.Lock()
+	defer n.executorStatisticsLock.Unlock()
+	n.executorStatistics[executorID] = stats
```
this is unexpected to see in **get**ExecutorStatistics. It's not clear why this side-effect here.

dkrotx reviewed (commented):
Multiple comments, the most common one - you need to split logic and infrastructure (retrieving/parsing).

function which performs:
- getData()
- modifyData()
- writeNewData()

is better than having some or all of them combined. Then function names will correspond more to what they're doing.

dkrotx commented on service/sharddistributor/store/etcd/executorstore/etcdstore.go:
```diff
@@ -136,6 +137,74 @@ func (s *executorStoreImpl) RecordHeartbeat(ctx context.Context, namespace, exec
+		return err
+	}
+
+	return nil
```
```suggestion
	return err
```

Theis-Mathiassen replied:
That should be changed now too

AndreasHolt committed:
chore: return error directly, reference ewma as smoothed load

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

Theis-Mathiassen committed:
chore: Removed no longer needed statistics from test, and magic number

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
Merge branch 'heartbeat-shard-statistics' of github.com:AndreasHolt/cadence into heartbeat-shard-statistics

Theis-Mathiassen committed:
chore: moved large switch case into separate function

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: refactored GetExecutorStatistics, fixed some concurrency problems, and ensure naming match functionality

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: moved statistics relevant data to separate struct

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: moved and renamed ewmasmoothload calculation

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: renamed mutable variable to something more descriptive, and made error handling clearer

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: returning error from applyStatistics as we now treat statistics as nescessary

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: responding on applyShardStatistics now returning error

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: continues best effort to update statistics, and return multiple errors

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: renamed recordShardStatistics in order to reflect what is happening, and extracted the actual writing of statistics out of it

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: moved the load check to calculate smoothload function

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: extract statistic update from etcdstore

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: added a timesource to namespaceShardToExecutor and ShardToExecutorCache, so we can control it with mock time in tests

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: refactored refreshExecutorState to adhere closer to MVC architecture

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

Theis-Mathiassen commented on service/sharddistributor/store/etcd/executorstore/etcdstore.go:
```diff
@@ -136,6 +137,74 @@ func (s *executorStoreImpl) RecordHeartbeat(ctx context.Context, namespace, exec
+	oldStats, err := s.shardCache.GetExecutorStatistics(ctx, namespace, executorID)
+	if err != nil {
+		return err
+	}
```
We only want to do this once before iterating over all shards for a executor, but i moved it a bit further down, I hope it helps: [3f0e893](https://github.com/cadence-workflow/cadence/commit/3f0e89323b1939853ebaf46c97ecaa82ecae43c5)

Theis-Mathiassen commented on service/sharddistributor/store/etcd/executorstore/etcdstore.go:
```diff
@@ -136,6 +137,74 @@ func (s *executorStoreImpl) RecordHeartbeat(ctx context.Context, namespace, exec
+	return nil
+}
+
+func (s *executorStoreImpl) recordShardStatistics(ctx context.Context, namespace, executorID string, reported map[string]*types.ShardStatusReport) error {
```
I renamed it to calcUpdatedStatistics, and returned the result instead of also applying them.
The change is in: [3f0e893](https://github.com/cadence-workflow/cadence/commit/3f0e89323b1939853ebaf46c97ecaa82ecae43c5)

Theis-Mathiassen committed:
chore: kept the lock while cloning, and moved right before ok created

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -97,6 +100,42 @@ func (n *namespaceShardToExecutor) GetExecutorModRevisionCmp() ([]clientv3.Cmp,
+	n.executorStatisticsLock.RUnlock()
+	if ok {
+		return cloned, nil
+	}
```
I was a bit unsure with how to deal with the unlocking, while keeping the code as simple as possible, i hope this is an acceptable solution: [dff7f07](https://github.com/cadence-workflow/cadence/commit/dff7f0751b98e5288015e9661209415f3307e415).

Theis-Mathiassen commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -124,14 +163,21 @@ func (n *namespaceShardToExecutor) nameSpaceRefreashLoop() {
+				case etcdkeys.ExecutorShardStatisticsKey:
+					n.handleExecutorStatisticsEvent(executorID, event)
+				case etcdkeys.ExecutorAssignedStateKey, etcdkeys.ExecutorMetadataKey:
 					shouldRefresh = true
```
It is moved

Theis-Mathiassen commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache_test.go:
```diff
@@ -128,13 +129,13 @@ func TestNamespaceShardToExecutor_Subscribe(t *testing.T) {
-// setupExecutorWithShards creates an executor in etcd with assigned shards and metadata
-func setupExecutorWithShards(t *testing.T, testCluster *testhelper.StoreTestCluster, executorID string, shards []string, metadata map[string]string) {
+// setupExecutorWithShards creates an executor in etcd with assigned shards, metadata, and optional statistics
+func setupExecutorWithShards(t *testing.T, testCluster *testhelper.StoreTestCluster, executorID string, shards []string, metadata map[string]string, stats map[string]etcdtypes.ShardStatistics) {
```
It was currently not needed in a test, so if it is needed I think we will call it explicitely.
[8ae5baa](https://github.com/cadence-workflow/cadence/commit/8ae5baa37a042fd4cda9c4211bda0ce278044402)

Theis-Mathiassen committed:
fix: changed the way we wait using timesource in tests

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache_test.go:
```diff
@@ -174,6 +180,71 @@ func verifyExecutorInState(t *testing.T, state map[*store.ShardOwner][]string, e
+	namespaceShardToExecutor, err := newNamespaceShardToExecutor(testCluster.EtcdPrefix, testCluster.Namespace, testCluster.Client, stopCh, logger)
+	assert.NoError(t, err)
+	namespaceShardToExecutor.Start(&sync.WaitGroup{})
+	time.Sleep(50 * time.Millisecond) // Give cache time to refresh
```
That should be resolved with these changes: [4625397](https://github.com/cadence-workflow/cadence/commit/462539776e86a2316f6a1a3fe94959524eaa4903)
This was another approach but block until seemed more elegant:
[eba5943](https://github.com/cadence-workflow/cadence/commit/eba5943fcbaec140da60d3cdd27a1603a14874ed)

Theis-Mathiassen committed:
fix: changed some time.Now to use mocktime source

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

Theis-Mathiassen committed:
fix: removed duplicate timesource

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

dkrotx reviewed (commented):
Good changes!
Let's proceed with making it more clear and moving algorithm-related code out store/etc tree.

dkrotx commented:
I feel like the commit message should be updated as well "Heartbeat shard statistics" points out to just statistics of heartbeats, but as far as I understand you introduce shard statistics and its smoothing in general. Delivering by heartbeat is minor here, as we know all the executor <-> SD comm is via heartbeats.

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

Theis-Mathiassen committed:
chore: moved CalculateSmoothedLoad out of executor store

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: added alias to metadata

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: refactored GetExecutorStatistics

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: added error logging statements to parse executor key

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: removed functions only used once

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: simplified return in RecordHeartbeat

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: refactored calcUpdatedStatistics

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: fixed duplicate error messages

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -214,49 +335,90 @@ func (n *namespaceShardToExecutor) refreshExecutorState(ctx context.Context) err
+	defer n.executorStatistics.lock.Unlock()
+	if len(stats) == 0 {
+		delete(n.executorStatistics.stats, executorID)
+		return
```
Yeah that is confusing, since these functions were only used in the function above, and were not too complex, i decided to remove them, and handle it in handleExecutorStatisticsEvent.
See [f7c9d77](https://github.com/cadence-workflow/cadence/commit/f7c9d77380499be8687ab5d323bcb164fb69fae8).

dkrotx commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -214,49 +339,84 @@ func (n *namespaceShardToExecutor) refreshExecutorState(ctx context.Context) err
-	return nil
+	n.executorStatistics.lock.Lock()
+	defer n.executorStatistics.lock.Unlock()
+	n.executorStatistics.stats[executorID] = cloneStatisticsMap(stats)
```
same as above. I would implement delete and assign as special methods managing locks themselves.
Then calling them becomes easier.

dkrotx commented on service/sharddistributor/store/etcd/executorstore/etcdstore.go:
```diff
@@ -722,57 +775,79 @@ func (s *executorStoreImpl) GetExecutor(ctx context.Context, namespace string, e
-				stats.SmoothedLoad = 0
-				stats.LastUpdateTime = etcdtypes.Time(now)
+				clonedOldOwnerStats := make(map[string]etcdtypes.ShardStatistics, len(oldOwnerStats))
+				maps.Copy(clonedOldOwnerStats, oldOwnerStats)
```
Actually, https://pkg.go.dev/maps#Clone should be a good fit.

Theis-Mathiassen committed:
chore: added logging to parseExecutorData

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: refactored executorData to be for a single executor

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: added some spacing

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
feat: added helper functions for namespaceExecutorStatistics

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: refactored introducing maps.clone

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: refactored switch in watch in namespaceshardcache

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
chore: made reason for deleting statistics more clear in handleExecutorStatisticsEvent

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -38,8 +39,63 @@ type namespaceShardToExecutor struct {
+	for _, kv := range resp.Kvs {
+		executorID, keyType, err := etcdkeys.ParseExecutorKey(etcdPrefix, namespace, string(kv.Key))
+		if err != nil {
+			continue
```
I do not think it is critical, at least in our load balancing, it is not ideal, but okay to just use the next most recent data. But agree we should add a warning: [38cf3d9](https://github.com/cadence-workflow/cadence/commit/38cf3d907f8999a4e18348765bd9ec377b8bac4c)

Theis-Mathiassen commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -102,6 +161,60 @@ func (n *namespaceShardToExecutor) GetExecutorModRevisionCmp() ([]clientv3.Cmp,
+func (n *namespaceShardToExecutor) readStats(executorID string) (map[string]etcdtypes.ShardStatistics, bool) {
+	n.executorStatistics.lock.RLock()
+	defer n.executorStatistics.lock.RUnlock()
+	stats, ok := n.executorStatistics.stats[executorID]
```
[09a4a40](https://github.com/cadence-workflow/cadence/commit/09a4a4061eba975b1067c8f74c7a3fe645d812f9)

dkrotx commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -214,49 +337,77 @@ func (n *namespaceShardToExecutor) refreshExecutorState(ctx context.Context) err
+	n.executorStatistics.lock.Lock()
+	defer n.executorStatistics.lock.Unlock()
+
+	n.applyParsedData(parsedData)
```
be consistent with naming. If one functions is called parseExecutorData(), then this one most probably should be `applyExecutorData` (I think it is assumet its parsed), or maybe `updateExecutorData()`.
Btw, sometimes that's a sign of artificial split of functions - when it's not easy to pick up the name.

dkrotx reviewed (approved):
I don't have strong objections anymore. Good job on addressing the comments!
Left some nit comments to address before landing (please squash the diff, or ask @jakobht to merge).

One thing to update (don't know if it possible at this stage) - there are multiple functions accessing executorStatistics and therefore knowing a lot about how update it (incl. if lock has to be taken). Try to make if abstract for outside users and just have update() or get() functions. This way:
- called shouldn't be aware of what is has to lock (it won't be possible to miss)
- you can easily change the structure without changing the code
- this code-block can be unit-tested separately, so when you build on top of it you already know you have something tested (and you can rely on it)
-- btw, I'd recommend using TDD in this case - writing test beforehand will define a clear   interface and you will immediately see how the usafe gonna look like.

Building with simple abstraction levels and combining them is a key for scalable development. Complex **intertwined** things are hard to make work reliably.

Theis-Mathiassen committed:
chore: refactored namespaceShardsCache

Signed-off-by: Theis Mathiassen <theis.mathiassen@gmail.com>

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

AndreasHolt committed:
fix: tests failing

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix: same fix but for shardcache

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
Merge remote-tracking branch 'origin/master' into heartbeat-shard-statistics

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

Theis-Mathiassen committed:
fix: test match functionality, and watch error not checked before processing events

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

Theis-Mathiassen committed:
fix: initial write of statistics would fail

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
fix: performance holding locks for less time and bug where executorStateChanges could skip statistics update

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
fix: properly return executor not found without wrapper to satisfy test

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
fix: correct handling of executor not found, and more useful error message

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

Theis-Mathiassen committed:
fix: test not building

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
chore: added nan check and made diff from main cleaner

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
feat: only write statistics if using greedy load balancing

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
fix: remove deleted executors from statistics

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

AndreasHolt committed:
Merge branch 'cadence-workflow:master' into heartbeat-shard-statistics

AndreasHolt committed:
Merge remote-tracking branch 'origin/master' into heartbeat-shard-statistics

AndreasHolt committed:
fix: executor verifies that it's the shard owner before overwriting map (to prevent multiple owners)

before this, there could be multiple owners of a shard temporarily, as
we didn't check cache. as a result, when the namespace state gets
flattened for GetState, the last apperance of that shards stats is the
only one that mattered.

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
test(shard-distributor): fix heartbeat shard statistics tests

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

Theis-Mathiassen committed:
Merge branch 'master' into heartbeat-shard-statistics

AndreasHolt committed:
fix: print the correct error

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
Merge branch 'heartbeat-shard-statistics' of github.com:AndreasHolt/cadence into heartbeat-shard-statistics

jakobht commented on service/sharddistributor/statistics/stats.go:
```diff
@@ -0,0 +1,25 @@
+package statistics
```
We need tests for this package

Theis-Mathiassen replied:
Added in [97ce863](https://github.com/cadence-workflow/cadence/pull/7431/commits/97ce86304c45a38b55df95cbdd3d03672af8420f)

jakobht commented on service/sharddistributor/statistics/stats.go:
```diff
@@ -0,0 +1,25 @@
+)
+
+func CalculateSmoothedLoad(prev, current float64, lastUpdate, now time.Time) float64 {
+	if math.IsNaN(current) || math.IsInf(current, 0) {
+		current = 0
+	}
+	if math.IsNaN(prev) || math.IsInf(prev, 0) {
+		prev = 0
+	}
```
I'd consider erroring out here - it would be really had to find a bug in the future when we just swallow the error.

Theis-Mathiassen replied:
That makes sense, added in [2fad64f](https://github.com/cadence-workflow/cadence/pull/7431/commits/2fad64f1988676f41c03fd418bffb00d7758bd45)

jakobht commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -42,22 +43,89 @@ type namespaceShardToExecutor struct {
+
+type namespaceExecutorStatistics struct {
+	lock  sync.RWMutex
+	stats map[string]map[string]etcdtypes.ShardStatistics
```
This is a quite complicated type, lets add a comment on what excatly is being mapped

Theis-Mathiassen replied:
It is, i added a comment: [5b0263b](https://github.com/cadence-workflow/cadence/pull/7431/commits/5b0263bfad8df517cb4c7e2a54296da896201311)

jakobht commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -42,22 +43,89 @@ type namespaceShardToExecutor struct {
+}
+
+type executorData struct {
+	assignedStates etcdtypes.AssignedState
```
Should this be assignedShards? Why do we need to save this again, isn't it redundant data?

Theis-Mathiassen replied:
The struct was constructed the way it was to match what we read from etcd in parseExecutorData function.
But you are right that it should be assignedShards instead of AssignedState, as that is the only field inside it we need.
I have simplified it a bit in: [753444d](https://github.com/cadence-workflow/cadence/pull/7431/commits/753444d469041b8f35dffd63e62e5228f77afc6a)

I am unsure what you mean with the redundant data, this struct is used to organize data between function calls in the refreshExecutorState function, and is dropped afterwards.
It was introduced in response to this [comment](https://github.com/cadence-workflow/cadence/pull/7431#discussion_r2610573302).

Let me know if this solves the issue :)

jakobht commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -42,22 +43,89 @@ type namespaceShardToExecutor struct {
+	assignedStates etcdtypes.AssignedState
+	metadata       map[string]string
+	statistics     map[string]etcdtypes.ShardStatistics
+	revisions      int64
```
```suggestion
	revision       int64
```

Theis-Mathiassen replied:
fixed in: [753444d](https://github.com/cadence-workflow/cadence/pull/7431/commits/753444d469041b8f35dffd63e62e5228f77afc6a)

jakobht commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -42,22 +43,89 @@ type namespaceShardToExecutor struct {
+	revisions      int64
+}
+
+func (n *namespaceShardToExecutor) parseExecutorData(resp *clientv3.GetResponse, etcdPrefix, namespace string) (map[string]executorData, error) {
```
Do we not already have a function doing this? This does not feel like it belongs in the cache?

Theis-Mathiassen replied:
There did not exist a function that did exactly what we needed, but there was a lot of the same functionality in GetState, and GetHeartbeat in etcdstore.go.

I have tried to create a common function in [243dbb9](https://github.com/cadence-workflow/cadence/pull/7431/commits/243dbb9758e30b1a3f265e2e8dfd91c682f0db3d) that could be used in all 3 cases, and put it in common/parse.go.

jakobht commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -42,22 +43,89 @@ type namespaceShardToExecutor struct {
+				return nil, fmt.Errorf("parse assigned state for %s: %w", executorID, err)
+			}
+			execData.assignedStates = assignedState
+			execData.revisions = kv.ModRevision
```
why do we exactly want this specific revision?

AndreasHolt replied:
The race it protects against is for example:
- writer A cache says shard-1 is unassigned
- writer B assigns shard-1 to executor-B
- writer A still thinks shard-1 is unassigned and assigns it to executor-A

We specifically keep the assigned_state key’s ModRevision because shard ownership is derived from assigned_state.
Then AssignShard later uses these cached revisions as transaction guards, so if any executor’s assignment state changed after the cache snapshot, the transaction fails instead of writing based on stale ownership info.

On a conflict, AssignShard retries after rereading ownership, so the next attempt is based on the latest assignment state.

jakobht commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -105,6 +173,63 @@ func (n *namespaceShardToExecutor) GetExecutorModRevisionCmp() ([]clientv3.Cmp,
+	n.executorStatistics.lock.Lock()
+	defer n.executorStatistics.lock.Unlock()
+
+	if _, ok := n.executorStatistics.stats[executorID]; ok {
```
Not sure I understand this - so we call refreash, but if we already have something in the cache we just ignore the response?

Theis-Mathiassen replied:
It was to avoid a situation where multiple executors would experience a cache miss, and both updating the cache one after the other, this should solve it, as the second executor sees that the cache now is populated and does not have to refresh it.

jakobht replied:
Lets write a comment that this is just an optimization - or maybe remove it.

AndreasHolt replied:
This helper is only meant to populate the cache on a cache miss by reading from etcd. The extra check is there to avoid overwriting a newer cached value if the watch loop populated the entry while this goroutine was still reading from etcd.

Example:
1. cache miss
2. this goroutine reads etcd
3. the watch loop populates the cache with a newer value
4. this goroutine acquires the lock later, after the etcd read
5. without the check, it would overwrite that newer cached value

We keep the guard for now, but have added a clarifying comment:

```golang
// The watch loop may have populated the cache while reading etcd.
// In that case, do not overwrite the newer cached value.
if _, ok := n.executorStatistics.stats[executorID]; ok {
      return nil
}
```

Refer to commit [7dccdc3](https://github.com/cadence-workflow/cadence/pull/7431/commits/7dccdc302199199a4fa61c87b5fffbc9c3f2d27b)

jakobht commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -105,6 +173,63 @@ func (n *namespaceShardToExecutor) GetExecutorModRevisionCmp() ([]clientv3.Cmp,
+		return nil
+	}
+
+	if len(resp.Kvs) > 0 {
```
I agree

AndreasHolt replied:
Agreed. I think the old name was misleading. This is not really a "force a refresh path", it's meant to be a population path on cache-miss population. 

we do the early return do avoid this cache miss race:
	1.	GetExecutorStatistics("exec-1") checks the cache.
	2.	No exec-1 entry exists.
	3.	It calls the refreshExecutorStatisticsCache function.
	4.	The function reads exec-1 stats from etcd.
	5.	Before it acquires the cache lock later, another goroutine or the etcdwatch loop may populate exec-1.
	6.	After acquiring the lock, the helper checks the cache again and returns without overwriting if the entry now exists.

I renamed the helper from refreshExecutorStatisticsCache to populateExecutorStatisticsCacheOnMiss to make that behavior explicit.

if it was meant to force a refresh, it would need revision tracking so an older etcd read cannot overwrite a newer value already applied.

Refer to commit [1cbd5fc](https://github.com/cadence-workflow/cadence/pull/7431/commits/1cbd5fc0d09b08b88018990000cd05a3d65803dd)

jakobht commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -276,49 +401,78 @@ func (n *namespaceShardToExecutor) refreshExecutorState(ctx context.Context) err
+	// Clear statistics to remove stale entries for deleted executors
+	n.executorStatistics.lock.Lock()
+	n.executorStatistics.stats = make(map[string]map[string]etcdtypes.ShardStatistics)
+	n.executorStatistics.lock.Unlock()
```
Is this needed? We remove all and then below we do a clean assignment. 

Also this introduces a race condition - for some period of time there is no statistics at all. This will cause cache misses and refreashs.

jakobht replied:
Also here we are mixing two locks with is always a red flag, do we need both locks? Can we do with one of them? Or maybe this should just be removed completely?

AndreasHolt replied:
> Is this needed? We remove all and then below we do a clean assignment.
> Also this introduces a race condition - for some period of time there is no statistics at all. This will cause
> cache misses and refreashs.

Agreed. Rather than clearing -> repopulating and introducing a race condition, we can build the stats map locally and swap it in once.

> Also here we are mixing two locks with is always a red flag, do we need both locks? Can we do with one of them? Or maybe this should just be removed completely?

For the lock concern, I agree that nesting locks is unnecessary. We still need to remove stale statistics for deleted executors during a full refresh, but we can do that without taking both locks at the same time.
We now build the refreshed ownership state and refreshed statistics state separately, then swap each under it's own lock. I'd keep the statistics lock separate for now since the statistics are also updated independently from statistics watch events, but avoid nested locking here.

Refer to commit
[4c7fcfd](https://github.com/cadence-workflow/cadence/pull/7431/commits/4c7fcfdf597c5429c959b5ac9666e5ac2448f6e6)

jakobht commented on service/sharddistributor/store/etcd/executorstore/shardcache/namespaceshardcache.go:
```diff
@@ -263,49 +382,73 @@ func (n *namespaceShardToExecutor) refreshExecutorState(ctx context.Context) err
+	return nil
+}
+
+func (n *namespaceShardToExecutor) applyExecutorData(data map[string]executorData) {
```
lets call the argument `executorData` or `executors` or something. `data` is very generic 
```suggestion
func (n *namespaceShardToExecutor) applyExecutorData(executorData map[string]executorData) {
```

AndreasHolt replied:
renamed to "executors" as I think it makes the most sense since map[string]executorData is a map of executor ID to executor data. 
So then the loop reads pretty naturally: "for executorID, executorData := range executors { ..." 

Commit [e5b33ef](https://github.com/cadence-workflow/cadence/pull/7431/commits/e5b33eff4eed2efaa20122bc893f1f3f2fcbbda4)

jakobht commented on service/sharddistributor/store/etcd/executorstore/shardcache/shardcache.go:
```diff
@@ -63,6 +64,14 @@ func (s *ShardToExecutorCache) GetShardOwner(ctx context.Context, namespace, sha
 	return namespaceShardToExecutor.GetShardOwner(ctx, shardID)
 }
 
+func (s *ShardToExecutorCache) GetExecutorStatistics(ctx context.Context, namespace, executorID string) (map[string]etcdtypes.ShardStatistics, error) {
```
We should have a test for this method

Theis-Mathiassen replied:
This has been added in [58f0098](https://github.com/cadence-workflow/cadence/pull/7431/commits/58f009811d00fc8296729e5704a062611172effe)

jakobht commented on service/sharddistributor/store/etcd/executorstore/etcdstore.go:
```diff
@@ -148,9 +151,84 @@ func (s *executorStoreImpl) RecordHeartbeat(ctx context.Context, namespace, exec
+	return []shardStatisticsUpdate{statsUpdate}, nil
+}
+
+func UpdateShardStatistic(shardID string, shardLoad float64, now time.Time, oldStats map[string]etcdtypes.ShardStatistics) etcdtypes.ShardStatistics {
```
I think this should be a private function.

Theis-Mathiassen replied:
Done in: [a435ddb](https://github.com/cadence-workflow/cadence/pull/7431/commits/a435ddb825be2d52add890f2ae33b6478cbd9e84)

jakobht commented on service/sharddistributor/store/etcd/executorstore/etcdstore.go:
```diff
@@ -864,94 +970,33 @@ func (s *executorStoreImpl) prepareShardStatisticsUpdates(ctx context.Context, n
 // applyShardStatisticsUpdates updates shard statistics.
-// Is intentionally made tolerant of failures since the data is telemetry only.
-func (s *executorStoreImpl) applyShardStatisticsUpdates(ctx context.Context, namespace string, updates []shardStatisticsUpdate) {
+func (s *executorStoreImpl) applyShardStatisticsUpdates(ctx context.Context, namespace string, updates []shardStatisticsUpdate) error {
```
The call side on line 449 is ignoring the error returned.

Theis-Mathiassen replied:
That is not ideal, it is resolved here in [09fa493](https://github.com/cadence-workflow/cadence/pull/7431/commits/09fa493e2f5f518820cff97613ffb12524951be0)

jakobht commented on service/sharddistributor/store/etcd/executorstore/etcdstore.go:
```diff
@@ -803,57 +881,85 @@ func (s *executorStoreImpl) GetExecutor(ctx context.Context, namespace string, e
+						if errors.Is(err, store.ErrExecutorNotFound) {
+							oldOwnerStats = make(map[string]etcdtypes.ShardStatistics)
+						} else {
+							return nil, err
```
Lets wrap this error

AndreasHolt replied:
Wrapped the error in [848803e](https://github.com/cadence-workflow/cadence/pull/7431/commits/848803e0291e44b21bbeb3b6cfeea40cd3487ce9)
Found some other adjacent errors in the same stats path that have also been wrapped in [9c8ef91](https://github.com/cadence-workflow/cadence/pull/7431/commits/9c8ef9101f84c847a064430c9677e1a201683ad6)

jakobht commented on service/sharddistributor/store/etcd/executorstore/etcdstore.go:
```diff
@@ -803,57 +881,85 @@ func (s *executorStoreImpl) GetExecutor(ctx context.Context, namespace string, e
+// This function calculates the necessary changes to shard statistics based on a new shard assignment plan.
+// It determines which shards have moved between executors, which are new. It then prepares a list of
+// update operations that will remove a moved shard's stats from its old owner and add them to its new owner, recording the time of the move.
 func (s *executorStoreImpl) prepareShardStatisticsUpdates(ctx context.Context, namespace string, newAssignments map[string]store.AssignedState) ([]shardStatisticsUpdate, error) {
```
I think this function works, but it's very hard to read. There are a lot of checks and variables to keep track of. I think it can be simplified quite a lot. 
In particular I think that the `existingShardFound` and the `if shardStatToMove.LastUpdateTime == etcdtypes.Time(time.Time{}) {` are very had to understand. 

Also doing a full clone of all stats of all shards for every shard makes this function quadratic. I think we should be able to avoid that.

AndreasHolt replied:
Agreed. Refactored to avoid the repeated map cloning and to make the flow more readable.

The new version handles the ownership cases directly. Shards already owned by the target are skipped and moved shards carry over previous stats when available.

Each affected executor’s stats map is loaded once and mutated in place.

Commit [11f10f4](https://github.com/cadence-workflow/cadence/commit/11f10f40de3feb3d5ae097a308418947976664fd)

jakobht reviewed (commented):
Generally looks good, but found a few issues we need to fix before merging.

Theis-Mathiassen committed:
fix: added error to CalculateSmoothedLoad

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
chore: added type explaining stats type

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
feat: added test for stats.go

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
chore: simplified executorData

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
chore: private function, updateShardStatistic

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
fix: handling previous ignored error

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
test: added test for GetExecutorStatistics

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
chore: wrap errors in prepareShardStatisticsUpdates

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
chore: additional error wrapping

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix, refactor: race condition in applyExecutorData. Remove nested
locking. Clean up into functions.

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
chore: rename data -> executors

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
refactor(sharddistributor): avoid repeated shard stats cloning and
simplify stats update

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
refactor(sharddistributor): avoid repeated shard stats cloning and
simplify stats update

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

Theis-Mathiassen committed:
chore: simplified lock in namespaceExecutorStatistics

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

AndreasHolt committed:
chore: rename refreshExecutorStatisticsCache ->
populateExecutorStatisticsCacheOnMiss for clarity

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

Theis-Mathiassen committed:
refactor: extracted functionality from getState, getHeartbeat, and parseExecutorData into 1 function ParseExecutorKVs

Signed-off-by: Theis Randeris Mathiassen <theis@tmathiassen.dk>

Theis-Mathiassen committed:
Merge branch 'heartbeat-shard-statistics' of github.com:AndreasHolt/cadence into heartbeat-shard-statistics

AndreasHolt committed:
Merge remote-tracking branch 'origin/master' into heartbeat-shard-statistics

AndreasHolt committed:
Merge branch 'master' into heartbeat-shard-statistics

AndreasHolt committed:
fix: format

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
fix(sharddistributor): remove unnecessary stats parse check

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>

AndreasHolt committed:
Merge branch 'master' into heartbeat-shard-statistics

AndreasHolt committed:
chore: add comment explaining guard in stats cache population on miss

Signed-off-by: Andreas Holt <6665487+AndreasHolt@users.noreply.github.com>
